torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -737,13 +496,165 @@ Stmt* LoopNest::lowerToStmt(Tensor* t) {
   return body;
 }
 
-void LoopNest::computeInline(Stmt* s) {
-  // TODO: check if `s` is a body of a loop
-  inlined_functions_.insert(stmt_to_tensor_.at(s)->function());
-}
+class FunctionInliner : public IRMutator {
+ public:
+  FunctionInliner(Store* producer)
+      : buf_(producer->buf()), producer_(producer) {}
+
+ protected:
+  // For the target function, insert the caller/callee pair into the replacement
+  // mapping.
+  const Expr* mutate(const FunctionCall* v) override {
+    Function* func = v->tensor()->function();
+    const Buf* buf = v->tensor()->buf();
+    // TODO: Support multiple-output functions
+    if (func->func_vars().size() != 1) {
+      throw unimplemented_lowering();
+    }
 
-void LoopNest::computeInlineWithRandom(Stmt* s) {
-  inlined_random_functions_.insert(stmt_to_tensor_.at(s)->function());
+    if (buf == buf_) {
+      // Insert the caller/callee pair into the mapping.
+      for (size_t i = 0; i < buf->ndim(); i++) {
+        const Var* func_callee_arg = dynamic_cast<const Var*>(func->arg(i));
+        const Expr* func_caller_param = v->param(i);
+        auto iter = inline_mapping_.find(func_callee_arg);
+        if (iter != inline_mapping_.end()) {
+          throw std::runtime_error(
+              "Duplicated variables: " + func_callee_arg->name_hint());
+        }
+        inline_mapping_[func_callee_arg] = func_caller_param;
+      }
+
+      // Call the actual replacement.
+      const Expr* body = producer_->value();
+      const Expr* result = body->accept_mutator(this);
+
+      // Remove the caller/callee relationship.
+      for (size_t i = 0; i < buf->ndim(); i++) {
+        const Var* func_callee_arg = dynamic_cast<const Var*>(func->arg(i));
+        auto iter = inline_mapping_.find(func_callee_arg);
+        if (iter == inline_mapping_.end()) {
+          throw std::runtime_error(
+              "Var already removed: " + func_callee_arg->name_hint());
+        }
+        inline_mapping_.erase(iter);
+      }
+      return result;
+    } else {
+      return IRMutator::mutate(v);
+    }
+  }
+
+  // Replace the target variable with the caller expressions.
+  const Expr* mutate(const Var* v) override {
+    auto iter = inline_mapping_.find(v);
```

Comment:
I cannot wrap my head around how this code works. Here is an example I'm trying to walk through:
```
a[i0] = i0*i0; 0 <= i0 < 10
b[j0] = a[j0-8]; 0 <= j0 < 2
```
When we lower to loopnest, it would look like:
```
for (i = 0; i < 10; i++)
  a[i] = i*i // (1)
for (j = 0; j < 2; j++)
  b[j] = a(j-8) // (2) FunctionCall
```
(I used `i0` in the original definition and `i` in the loopnest to emphasize that those are two different `Var`s).

Now, as we run `FunctionInliner`, when we hit the `a(j-8)` function call, it would add a mapping `i0` -> `j-8`. Then, we're replacing `a(j-8)` with the value of store (1), which is `i*i`. Next step is to replace all variables according to our map, but there is not `i` in our map, only `i0` and thus it will not change anything. So, I would expect that after inlining we're getting something like:
```
for (i = 0; i < 10; i++)
  a[i] = i*i
for (j = 0; j < 2; j++)
  b[j] = i*i
```
Which is obviously incorrect. What did I miss in the process? One suspicion I have is that when we create a loop nest for `a` we're actually using the same variable as in the definition of the function, which means that in this example `i0` and `i` would actually be the same. However, this will not be true after we, say, split the loop of `a` - we would have new variables there. How would the new implementation handle that case? Do we have a test that checks that this works correctly?

Reply:
This feels like a weird example, without inlining that first program will crash when it indexes a[-8]. Is that what you meant?

I think the part you're missing is when we var substitute the params of the function call with the loop vars, I've verified that if a is not an output buf this works:
```
Tensor* a =
    Compute("a", {{M, "i"}}, [&](const VarHandle& i) { return i * i; });
Tensor* b = Compute("b", {{2, "j"}}, [&](const VarHandle& j) {
  return a->call(j - ExprHandle(8));
});

LoopNest loop({b});
loop.computeInline(a->buf());
loop.prepareForCodegen();
Stmt* s = IRSimplifier::simplify(loop.root_stmt());
```
prints:
```
{
  for (int j = 0; j < 2; j++) {
    b[j] = (j - 8) * (j - 8);
  }
}
```

If `a` is an output buf then the inline will fail.

Reply:
Sorry, I meant `b[j0] = a[j0+8]; 0 <= j0 < 2`.

Below is the test illustrating the problem:
```
  {
    KernelScope kernel_scope;
    Tensor* a =
        Compute("a", {{18, "i"}}, [&](const VarHandle& i) { return i * i; });
    Tensor* b = Compute("b", {{2, "j"}}, [&](const VarHandle& j) {
      return a->call(j + ExprHandle(8));
    });

    LoopNest loop({b});
    For* i_outer;
    For* i_inner;

    LoopNest l({b});
    std::vector<For*> loops = l.getLoopStmtsFor(a);
    std::cerr << "*** 1. Original:\n" << *l.root_stmt() << "\n";
    l.splitWithMask(loops[0], 4, &i_outer, &i_inner);
    std::cerr << "*** 2. After splitWithMask:\n" << *l.root_stmt() << "\n";
    l.prepareForCodegen();
    std::cerr << "*** 3. Final:\n" << *l.root_stmt() << "\n";
  }
  {
    KernelScope kernel_scope;
    Tensor* a =
        Compute("a", {{18, "i"}}, [&](const VarHandle& i) { return i * i; });
    Tensor* b = Compute("b", {{2, "j"}}, [&](const VarHandle& j) {
      return a->call(j + ExprHandle(8));
    });

    LoopNest loop({b});
    For* i_outer;
    For* i_inner;

    LoopNest l({b});
    std::vector<For*> loops = l.getLoopStmtsFor(a);
    std::cerr << "*** 1. Original:\n" << *l.root_stmt() << "\n";
    l.computeInline(a->buf());
    std::cerr << "*** 2. After computeInline:\n" << *l.root_stmt() << "\n";
    l.prepareForCodegen();
    std::cerr << "*** 3. Final:\n" << *l.root_stmt() << "\n";
  }
  {
    KernelScope kernel_scope;
    Tensor* a =
        Compute("a", {{18, "i"}}, [&](const VarHandle& i) { return i * i; });
    Tensor* b = Compute("b", {{2, "j"}}, [&](const VarHandle& j) {
      return a->call(j + ExprHandle(8));
    });

    LoopNest loop({b});
    For* i_outer;
    For* i_inner;

    LoopNest l({b});
    std::vector<For*> loops = l.getLoopStmtsFor(a);
    std::cerr << "*** 1. Original:\n" << *l.root_stmt() << "\n";
    l.splitWithMask(loops[0], 4, &i_outer, &i_inner);
    std::cerr << "*** 2. After splitWithMask:\n" << *l.root_stmt() << "\n";
    l.computeInline(a->buf()); // <<<THIS WILL CRASH
    std::cerr << "*** 3. After computeInline:\n" << *l.root_stmt() << "\n";
    l.prepareForCodegen();
    std::cerr << "*** 4. Final:\n" << *l.root_stmt() << "\n";
  }
```

The output of this test is:
```
*** 1. Original:
{
  for (int i = 0; i < 18; i++) {
    a[i] = i * i;
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 2. After splitWithMask:
{
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (i_outer * 4 + i_inner<18 ? 1 : 0) {
        a[i_outer * 4 + i_inner] = (i_outer * 4 + i_inner) * (i_outer * 4 + i_inner);
      }
    }
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 3. Final:
{
  Allocate(a, int, {18});
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (i_outer * 4 + i_inner<18 ? 1 : 0) {
        a[i_outer * 4 + i_inner] = (i_outer * 4 + i_inner) * (i_outer * 4 + i_inner);
      }
    }
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a[j + 8];
  }
  Free(a);
}


*** 1. Original:
{
  for (int i = 0; i < 18; i++) {
    a[i] = i * i;
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 2. After computeInline:
{
  for (int j = 0; j < 2; j++) {
    b[j] = (j + 8) * (j + 8);
  }
}
*** 3. Final:
{
  for (int j = 0; j < 2; j++) {
    b[j] = (j + 8) * (j + 8);
  }
}


*** 1. Original:
{
  for (int i = 0; i < 18; i++) {
    a[i] = i * i;
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 2. After splitWithMask:
{
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (i_outer * 4 + i_inner<18 ? 1 : 0) {
        a[i_outer * 4 + i_inner] = (i_outer * 4 + i_inner) * (i_outer * 4 + i_inner);
      }
    }
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 3. After computeInline:
{
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (!i_outer * 4 + i_inner<18 ? 1 : 0) Segmentation fault (core dumped)
```

Reply:
Thanks for taking the time to write the example code that makes it way easier.

So the crash is an unrelated bug thats easy to fix, but once it is fixed I see the issue you're referring to, we print out something like:
```
  for (int j = 0; j < 2; j++) {
    b[j] = (i_outer * 4 + i_inner) * (i_outer * 4 + i_inner);
  }
```

Which is bad, obviously.

I hacked up a fix for this last night and it looks good, we're now producing this output (after a small addition to the simplifier):
```
*** 1. Original:
{
  for (int i = 0; i < 18; i++) {
    a[i] = i * i;
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 2. After splitWithMask:
{
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (i_outer * 4 + i_inner<18 ? 1 : 0) {
        a[i_outer * 4 + i_inner] = (i_outer * 4 + i_inner) * (i_outer * 4 + i_inner);
      }
    }
  }
  for (int j = 0; j < 2; j++) {
    b[j] = a(j + 8);
  }
}
*** 3. After computeInline:
{
  for (int i_outer = 0; i_outer < (((18 - 0) + 4) - 1) / 4; i_outer++) {
    for (int i_inner = 0; i_inner < 4; i_inner++) {
      if (i_outer * 4 + i_inner<18 ? 1 : 0) {
{
        }      }
    }
  }
  for (int j = 0; j < 2; j++) {
    b[j] = (j + 8) * (j + 8);
  }
}
*** 4. Final:
{
  for (int j = 0; j < 2; j++) {
    b[j] = (j + 8) * (j + 8);
  }
}
```

I need a bit of time to clean it up, and I'd like to add a bunch of tests in this kind of style.